test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetUnnamedPackage.java

```diff
@@ -0,0 +1,96 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8284037
+ * @summary Snippet-files subdirectory not automatically detected when in unnamed package
+ * @library /tools/lib ../../lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.main
+ *          jdk.javadoc/jdk.javadoc.internal.tool
+ * @build javadoc.tester.* toolbox.ToolBox toolbox.ModuleBuilder builder.ClassBuilder
+ * @run main TestSnippetUnnamedPackage
+ */
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestSnippetUnnamedPackage extends SnippetTester {
+
+    public static void main(String... args) throws Exception {
+        new TestSnippetUnnamedPackage().runTests();
+    }
+
+    @Test
+    public void testNoSourcePath(Path base) throws IOException {
+        test(base, false);
+    }
+
+    @Test
+    public void testSourcePath(Path base) throws IOException {
+        test(base, true);
+    }
+
+    void test(Path base, boolean useSourcePath) throws IOException {
+        Path src = base.resolve("src");
+        tb.writeJavaFiles(src,
+                """
+                        /**
+                         * Comment.
+                         * Before.
+                         * {@snippet class=S}
+                         * After.
+                         */
+                        public class C {
+                          private C() { }
+                        }
+                        """);
+        tb.writeFile(src.resolve("snippet-files").resolve("S.java"),
+        "public class S { }");
+
+        var args = new ArrayList<String>();
+        args.addAll(List.of("-d", base.resolve("out").toString()));
+        if (useSourcePath) {
+            args.addAll(List.of("--source-path", src.toString()));
+        }
+        args.add(src.resolve("C.java").toString());
+
+        javadoc(args.toArray(String[]::new));
+        checkExit(useSourcePath ? Exit.OK : Exit.ERROR);
+        checkOutput(Output.OUT, !useSourcePath,
+                "C.java:4: error: file not found on source path or snippet path: S.java");
```

Comment:
I must admit I don't understand the rules for snippet discovery with or without --source-path argument. Would the same behavior be expcected for code in a named package?

Reply:
Well, the correct behavior is the same for both named and unnamed packages, but the bug was specifically about the use of the unnamed package.  

The correct behavior is that snippet files are first looked up on the source path (`--source-path` and friends), in the appropriate `snippet-files` subdirectory relative to the enclosing package, and if not found there, the snippet files are looked for on the snippet path (`--snippet-path`).  (( _An interesting corollary is that the `snippet-files` subdirectory does not need to be a direct subdirectory of the directory containing the snippet tag: it just has to exist with that name somewhere on the path. This could maybe be used by users and IDEs to their advantage._ ))

The underlying javac "rule" is that files are either given on the command-line or looked up on some search path. Snippet files can't be specified on the command-line; therefore, they must be found on a search path: in this case, either the source path or the snippet path.

The bug fixed in this issue was related to creating the name to give to the file manager. For the file manager, the unnamed package is represented by an empty string. We were (incorrectly) using the strings `<Unnamed>`. oops.